File: valofosho/BOJ/boj1406.py
"""
L: idx - 1 if idx == 0 > still
D: idx + 1 if idx == len > still
B: delete stack[idx], if idx == 0 > still
P $: $ 를 커서 왼쪽에 추가 -> insert $ idx
a b c d e
1 -> cursor = 5 a b c d e |
L -> cursor = 4 a b c d | e
L -> cursor = 3 a b c | d e
B -> cursor = 2 a b | d e
P -> cursor = 3 a b P | d e
"""



import sys
input = sys.stdin.readline
# 두 개의 좌, 우 스택을 활용한 풀이
# 커서를 기준으로 좌, 우 나누어서 활용
left = list(input().strip())
right = list()
N = int(input())
command = list(input().split() for _ in range(N))
for cmd in command:
    # L 연산은 커서를 왼쪽으로 옮긴다.
    # 즉, Left의 마지막 요소를 Right로 옮긴다
    if cmd[0] == "L":
        if left:
            right.append(left.pop())
    # R 연산은 커서를 왼쪽으로 옮긴다.
    # 즉, Rightt의 마지막 요소를 Left로 옮긴다
    elif cmd[0] == "D":
        if right:
            left.append(right.pop())
    # B 연산은 커서 왼쪽의 문자를 지운다
    elif cmd[0] == "B":
        if left:
            left.pop()
    # 커서 왼쪽에 $를 삽입
    else:
        left.append(cmd[1])
# 스택은 LIFO 구조이므로 Reverse
r = list(reversed(right))
print(''.join(left) + ''.join(r))




# 시간 초과로 사망
# def L(head):
#     if head != 0:
#         head -= 1   
#     return head

# def D(head, end):
#     if head < end:
#         head += 1
#     return head

# def B(head, end):
#     if head != 0:
#         stack.pop(head-1)
#         head -= 1
#         end -= 1
#     return head, end

# def P(head, value, end):
#     stack.insert(head, value)
#     head += 1
#     end += 1
#     return head, end

# import sys
# input = sys.stdin.readline

# stack = list(input().strip())
# N = int(input())
# command = list(input().split() for _ in range(N))
# cnt = len(stack)
# head = cnt
# for cmd in command:
#     if cmd[0] == "L":
#         head = L(head)
#     elif cmd[0] == "D":
#         head = D(head, cnt)
#     elif cmd[0] == "B":
#         head ,cnt = B(head, cnt)
#     else:
#         head, cnt = P(head, cmd[1], cnt)

# print(''.join(stack))

File: aeongseo/BOJ/CPP/boj1406.cpp
/*** 1406. 에디터 ***/

#include<iostream>
#include<string>
#include<algorithm>
#include<stack>
using namespace std;

int main() {
	stack<char> left, right;	// 커서 기준 왼쪽 문자와 오른쪽 문자 저장할 stack
	string txt, order;
	int cnt;
	cin >> txt >> cnt;

	for (int i = 0; i < txt.size(); i++) {	//입력받은 문자열 left에 모두 저장
		left.push(txt[i]);
	}
	for (int j = 0; j < cnt; j++) {	// 
		cin >> order;
		if (order == "L") {
			if (left.size() == 0) continue;	// left 스택이 비어있으면 아무것도 안함
			right.push(left.top());	// left의 top의 값을 오른쪽으로 이동 (커서 왼쪽으로 이동)
			left.pop();
		}
		else if (order == "D") {
			if (right.size() == 0) continue;	// right 스택이 비어있으면 아무것도 안함
			left.push(right.top());	// right의 top의 값을 왼쪽으로 이동 (커서 오른쪽으로 이동)
			right.pop();
		}
		else if (order == "B") {
			if (left.size() == 0) continue;	// left 스택이 비어있으면 아무것도 안함
			left.pop();	// left의 top pop
		}
		else if (order == "P") {
			char x;
			cin >> x;	// 추가할 값 입력받음
			left.push(x);	// left에 push
		}
	}
	string ans;
	while (!left.empty()) {	// left 스택이 빌 때까지
		ans.push_back(left.top());	// ans 에 값 추가
		left.pop();
	}
	reverse(ans.begin(), ans.end());	// ans reverse

	while (!right.empty()) {	// right 스택 빌 때까지
		ans.push_back(right.top());	// ans에 값 추가
		right.pop();
	}
	
	cout << ans;
	
	return 0;
}